src/persist-txn/src/txn_cache.rs

```diff
@@ -1049,127 +1060,145 @@ mod tests {
+        // Revisit when
+        // https://github.com/MaterializeInc/materialize/issues/25992 is fixed,
+        // it's unclear how to encode the register timestamp in a forget.
+        c.push_register(d0, 8, -1, 8);
+        testcase(&mut c, 8, d0, ds(None, 8, 9), EmitLogicalProgress(9));
+
+        // ts 9 (not registered, not written). This ReadDataTo would block until
+        // the write happens at ts 10.
+        testcase(&mut c, 9, d0, ds(None, 9, 10), ReadDataTo(10));
+
+        // ts 10 (written directly)
+        testcase(&mut c, 10, d0, ds(None, 10, 11), ReadDataTo(11));
+
+        // ts 11 (not registered, not written) This ReadDataTo would block until
+        // the register happens at 12.
+        testcase(&mut c, 11, d0, ds(None, 11, 12), ReadDataTo(12));
+
+        // ts 12 (registered, previously forgotten)
+        c.push_register(d0, 12, 1, 12);
+        testcase(&mut c, 12, d0, ds(None, 12, 13), EmitLogicalProgress(13));
+
+        // ts 13 (forgotten, registered at preceding ts)
+        // Revisit when
+        // https://github.com/MaterializeInc/materialize/issues/25992 is fixed,
+        // it's unclear how to encode the register timestamp in a forget.
+        c.push_register(d0, 13, -1, 13);
+        testcase(&mut c, 13, d0, ds(None, 13, 14), EmitLogicalProgress(14));
+
+        // Now that we have more history, some of the old answers change! In
+        // particular, we have more information on unapplied writes, empty
+        // times, and can ReadDataTo much later times.
+
+        assert_eq!(c.data_snapshot(d0, 0), ds(None, 0, 4));
+        assert_eq!(c.data_snapshot(d0, 1), ds(None, 1, 4));
+        assert_eq!(c.data_snapshot(d0, 2), ds(None, 2, 4));
+        assert_eq!(c.data_snapshot(d0, 3), ds(None, 3, 4));
+        assert_eq!(c.data_snapshot(d0, 4), ds(None, 4, 5));
+        assert_eq!(c.data_snapshot(d0, 5), ds(None, 5, 7));
+        assert_eq!(c.data_snapshot(d0, 6), ds(None, 6, 7));
+        assert_eq!(c.data_snapshot(d0, 7), ds(None, 7, 14));
+        assert_eq!(c.data_snapshot(d0, 8), ds(None, 8, 14));
+        assert_eq!(c.data_snapshot(d0, 9), ds(None, 9, 14));
+        assert_eq!(c.data_snapshot(d0, 10), ds(None, 10, 14));
+        assert_eq!(c.data_snapshot(d0, 11), ds(None, 11, 14));
+        assert_eq!(c.data_snapshot(d0, 12), ds(None, 12, 14));
+        assert_eq!(c.data_snapshot(d0, 13), ds(None, 13, 14));
+
+        assert_eq!(c.data_listen_next(&d0, 0), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 1), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 2), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 3), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 4), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 5), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 6), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 7), ReadDataTo(8));
+        assert_eq!(c.data_listen_next(&d0, 8), EmitLogicalProgress(14));
```

Comment:
Wait, isn't this wrong? The shard was written to directly at `10`, and this would essentially lose that write, no?

Reply:
Oops, yes great point. This should be `EmitLogicalProgress(9)` and then we should be returning a bunch of `ReadDataTo(12)`.

Reply:
Pretty worrisome that nothing in CI, including nightlies, seemed to catch this?

Reply:
I'm not sure, is this tested anywhere outside of this unit test?

Reply:
I think in practice we're not forgetting and re-registering many shards and we aren't asking about times in the past with `data_listen_next` which is why this doesn't blow up anywhere.

Reply:
My most recent commit should fix this though.

Reply:
FWIW here's a repro of this test case:

Note timestamp `t_n` is ordered by the value of `n`.

1. Register shard at `t_1`.
2. Forget shard at `t_2`.
3. Write directly to shard at `t_4`.
4. Register shard at `t_5`.
5. Call `data_listen_next` at `t_3`. This should return at least `ReadDataTo(t_4)` and not `EmitlogicalProgress(t_4)`.

Reply:
Good news! `cargo stress -p mz-persist-txn --lib` got this failure at the commit with the bug (after 1929 attempts). Haven't checked yet what happens at the commit with the fix

```
thread 'txns::tests::stress_correctness' panicked at src/persist-txn/src/txns.rs:1439:21:
assertion `left == right` failed
  left: [("w0s12", 91, 1), ("w0s33", 91, 1), ("w0s36", 91, 1), ("w0s40", 91, 1), ("w0s45", 91, 1), ("w0s55", 91, 1), ("w0s59", 93, 1), ("w1s20", 91, 1), ("w1s39", 91, 1), ("w1s40", 91, 1), ("w1s51", 91, 1), ("w1s68", 91, 1), ("w1s71", 91, 1), ("w1s72", 91, 1)]
 right: [("w0s12", 91, 1), ("w0s33", 91, 1), ("w0s36", 91, 1), ("w0s40", 91, 1), ("w0s45", 91, 1), ("w0s55", 91, 1), ("w0s59", 93, 1), ("w1s20", 91, 1), ("w1s39", 91, 1), ("w1s40", 91, 1), ("w1s51", 91, 1), ("w1s68", 91, 1), ("w1s71", 91, 1), ("w1s72", 91, 1), ("w1s98", 107, 1)]
```

Reply:
Yup, fixed at the latest commit!

```
20129 runs so far, 0 failures, over 3724s
```